src/query_graph/mod.rs

```diff
@@ -695,11 +696,56 @@ impl QueryGraph {
         };
     }
 
+    /// Returns a selection set that can be used as a key for the given type, and that can be
+    /// entirely resolved in the same subgraph. Returns None if such a key does not exist for the
+    /// given type.
     pub(crate) fn get_locally_satisfiable_key(
         &self,
-        _node: NodeIndex,
+        node_index: NodeIndex,
     ) -> Result<Option<NormalizedSelectionSet>, FederationError> {
-        todo!()
+        let node = self.node_weight(node_index)?;
+        let type_name = match &node.type_ {
+            QueryGraphNodeType::SchemaType(ty) => {
+                CompositeTypeDefinitionPosition::try_from(ty.clone())?
+            }
+            QueryGraphNodeType::FederatedRootType(_) => {
+                return Err(FederationError::internal(format!(
+                    "get_locally_satisfiable_key must be called on a composite type, got {}",
+                    node.type_
+                )));
+            }
+        };
+        let schema = self.schema_by_source(&node.source)?;
+        let Some(metadata) = schema.subgraph_metadata() else {
+            return Err(FederationError::internal(format!(
+                "Missing metadata on subgraph {}",
```

Comment:
Have we been keeping the error messages exactly the same in the port, or have we been changing them to be more aligned with Rust error messages? This one makes sense to me though.

Reply:
typically the same. the intentional difference here is to write "subgraph metadata" instead of "federation metadata" to align with the new name of the rust struct. i updated it to invent less new stuff in b139d136e24ac305110814779511b1960ad96bcd :)